proptest-state-machine/src/iterative_runner.rs

```diff
@@ -0,0 +1,861 @@
+    }
+
+    S::teardown(sut_state, mod_state)
+}
+
+pub trait IterativeModelStateMachine {
+    /// Model state used to drive the generation of the transitions
+    type ModelState: Clone + Debug;
+    /// Transition generated by this model
+    type Transition: Clone + Debug;
+    /// Concrete SUT state, which could be consulted to further sharpern
+    /// the state transition generation
+    type SutState;
+
+    /// Initial state of the model
+    fn init_state() -> BoxedStrategy<Self::ModelState>;
+
+    /// Generate transition based on the state of the model. System under
+    /// test can be consulted here as well to reduce number of generated
+    /// transitions
+    fn transition(
+        mod_state: &Self::ModelState,
+        sut_state: &Self::SutState,
+    ) -> BoxedStrategy<Self::Transition>;
+
+    /// Apply a transition to the model
+    fn apply(mod_state: Self::ModelState, transition: &Self::Transition) -> Self::ModelState;
+
+    /// Pre-conditions may be specified to control which transitions are valid
+    /// from the current state. If not overridden, this allows any transition.
+    /// The pre-conditions are checked in the generated transitions and during
+    /// shrinking. System under test can be consulted in both phases to discard
+    /// transitions.
+    fn preconditions(
+        mod_state: &Self::ModelState,
+        sut_state: &Self::SutState,
+        transition: &Self::Transition,
+    ) -> bool {
+        // This is to avoid `unused_variables` warning
+        let _ = (mod_state, sut_state, transition);
+        true
+    }
+}
+
+/// SUT state machine that relies on a machine model
+pub trait IterativeSutStateMachine {
+    /// The concrete state, that is the system under test (SUT).
+    type SutState;
+    /// Transition generated by the model
+    type Transition: Clone + Debug;
+    /// Abstract machine model
+    type ModelState: Clone + Debug;
+
+    /// Initialize the state of SUT.
+    ///
+    /// If the reference state machine is generated from a non-constant
+    /// strategy, ensure to use it to initialize the SUT to a corresponding
+    /// state.
+    fn init_test(ref_state: &Self::ModelState) -> Self::SutState;
+
+    /// Apply a transition in the SUT state and check post-conditions.
+    /// The post-conditions are properties of your state machine that you want
+    /// to assert.
+    ///
+    /// Note that the `mod_state` is the state *after* this `transition` is
+    /// applied. You can use it to compare it with your SUT after you apply
+    /// the transition.
+    fn apply(
+        sut_state: Self::SutState,
+        mod_state: &Self::ModelState,
+        transition: Self::Transition,
+    ) -> Self::SutState;
+
+    /// Check some invariant on the SUT state after every transition.
+    fn check_invariants(sut_state: &Self::SutState, mod_state: &Self::ModelState) {
+        // This is to avoid `unused_variables` warning
+        let _ = (sut_state, mod_state);
+    }
```

Comment:
To be honest given that Model in IterativeRunner is a less formulated and strict, I would prefer to keep this API as is, as it provides flexibility. For example for actual problem where I needed to use iterative runner the model is implementing both the simplified (not complete) model of the system under test, and also controls the external resources that system under test interacts with. For example the model may not be aware that system under test is issuing the calls towards the external subsystems (because the model is not complete), but system under test would be able to get this data from the model in the apply call. Somewhat similar in the check_invariants API depending on how the model is used, there might be desire to check that state of the model and state of the system under test are in fact in sync here and invariant holds based on the state of the model.

I understand that this explanation is a bit vague, but one of the intents of the IterativeRunner is to provide flexibility that otherwise can't be provided by the existing SequentialRunner.